plasmapy/simulation/particle_tracker/save_routines.py

```diff
@@ -0,0 +1,215 @@
+"""
+
+__all__ = [
+    "AbstractSaveRoutine",
+    "DoNotSaveSaveRoutine",
+    "SaveOnceOnCompletion",
+    "IntervalSaveRoutine",
+]
+
+from abc import ABC, abstractmethod
+from pathlib import Path
+
+import astropy.units as u
+import h5py
+
+
+class AbstractSaveRoutine(ABC):
+    """Abstract base class containing the necessary methods for a
+    `~plasmapy.simulation.particle_tracker.particle_tracker.ParticleTracker` save routine.
+
+    The save routine class is responsible for defining the conditions and hooks
+    for saving.
+
+    Parameters
+    ----------
+    output_directory : `~pathlib.Path`, optional
+        Output for objects that are saved to disk. If a directory is not specified
+        then a memory save routine is used.
+
+
+    Notes
+    -----
+    After every push, the `post_push_hook` method is called with the
+    respective `~plasmapy.simulation.particle_tracker.particle_tracker.ParticleTracker` object passed as a parameter.
+    Then, the hook calls `save_now` to determine whether or not the simulation state should be saved.
+    """
+
+    def __init__(self, output_directory: Path | None = None) -> None:
+        self.output_directory = output_directory
+
+        self._results = {}
+        self._quantities = {
+            "x": (u.m, "dataset"),
+            "v": (u.m / u.s, "dataset"),
+        }
+
+        self._particle_tracker = None
+
+    @property
+    def tracker(self):
+        """Return the `ParticleTracker` object for this stop condition."""
+        return self._particle_tracker
+
+    @tracker.setter
+    def tracker(self, particle_tracker) -> None:
+        self._particle_tracker = particle_tracker
+
+    @property
+    @abstractmethod
+    def require_synchronized_dt(self) -> bool:
+        """Return if this save routine requires a synchronized time step."""
+        ...
+
+    @property
+    @abstractmethod
+    def save_now(self) -> bool:
+        """Determine if to save on the current push step."""
+        ...
+
+    @property
+    def results(self) -> dict[str, u.Quantity]:
+        """Return the results of the simulation.
+        The quantities returned depend on those defined in the body of the save routine.
+        """
+        return self._apply_units_to_results()
+
+    def save(self) -> None:
+        """Save the current state of the simulation to memory.
+
+        If an output directory is specified then the state will also be saved
+        to the disk.
+        """
+
+        self._save_to_memory()
+
+        if self.output_directory is not None:
+            self._save_to_disk()
+
+    def _save_to_disk(self) -> None:
+        """Save a hdf5 file containing simulation positions and velocities."""
+
+        path = self.output_directory / f"{self.tracker.iteration_number}.hdf5"
+
+        with h5py.File(path, "w") as output_file:
+            for key, (_units, data_type) in self._quantities.items():  # noqa: B007
+                match data_type:
```

Comment:
Wow, somehow I completely missed that Python added this switch statement architecture in 3.10...